DM worker panic when OpenAPI GET /api/v1/tasks/{task-name}/sources/{source-name}/migrate_targets

### What did you do?

curl -X 'GET'   'http://10.2.8.3:8561/api/v1/tasks/task-1/sources/mysql-01/migrate_targets'   -H 'accept: application/json'

### What did you expect to see?

_No response_

### What did you see instead?

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x2d7c85d]

goroutine 223775 [running]:
github.com/pingcap/tidb/ddl/schematracker.(*InfoStore).AllSchemaNames(...)
        github.com/pingcap/tidb@v1.1.0-beta.0.20230412065101-c24a1cda6d26/ddl/schematracker/info_store.go:132
github.com/pingcap/tiflow/dm/pkg/schema.(*Tracker).AllSchemas(...)
        github.com/pingcap/tiflow/dm/pkg/schema/tracker.go:236
github.com/pingcap/tiflow/dm/syncer.(*Syncer).listMigrateTargets(0xc00207a900, 0xc002da1500)
        github.com/pingcap/tiflow/dm/syncer/schema.go:183 +0x5d
github.com/pingcap/tiflow/dm/syncer.(*Syncer).OperateSchema(0xc00207a900, {0x3faab90?, 0xc00056ed80}, 0xc002da1500)
        github.com/pingcap/tiflow/dm/syncer/schema.go:49 +0x185
github.com/pingcap/tiflow/dm/worker.(*SubTask).OperateSchema(0xc000d78460, {0x3faab90, 0xc00056ed80}, 0xc002da1500)
        github.com/pingcap/tiflow/dm/worker/subtask.go:678 +0x24c
github.com/pingcap/tiflow/dm/worker.(*SourceWorker).OperateSchema(0xc001346140, {0x3faab90, 0xc00056ed80}, 0xc002da1500)
        github.com/pingcap/tiflow/dm/worker/source_worker.go:1046 +0x1ae
github.com/pingcap/tiflow/dm/worker.(*Server).OperateSchema(0xc00169c120, {0x3faab90, 0xc00056ed80}, 0xc002da1500)
        github.com/pingcap/tiflow/dm/worker/server.go:848 +0x4a7
github.com/pingcap/tiflow/dm/pb._Worker_OperateSchema_Handler({0x35e11e0?, 0xc00169c120}, {0x3faab90, 0xc00056ed80}, 0xc0012df570, 0x0)
        github.com/pingcap/tiflow/dm/pb/dmworker.pb.go:3973 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc001834000, {0x3fcc3a0, 0xc000c321a0}, 0xc001a0a360, 0xc000d7aa50, 0x63e9ad0, 0x0)
        google.golang.org/grpc@v1.50.1/server.go:1340 +0xd23
google.golang.org/grpc.(*Server).handleStream(0xc001834000, {0x3fcc3a0, 0xc000c321a0}, 0xc001a0a360, 0x0)
        google.golang.org/grpc@v1.50.1/server.go:1713 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.50.1/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.50.1/server.go:963 +0x28a

### Versions of the cluster

v6.5.4

### current status of DM cluster (execute `query-status <task-name>` in dmctl)

```console
(paste current status of DM cluster here)
```

--- a/dm/tests/openapi/run.sh
+++ b/dm/tests/openapi/run.sh
@@ -752,6 +752,7 @@ function test_start_task_with_condition() {
 	run_sql_source2 "CREATE TABLE openapi.t2(i TINYINT, j INT UNIQUE KEY);"
 
 	openapi_task_check "create_incremental_task_with_gtid_success" $task_name "" "" "" "" "" ""
+	curl http://127.0.0.1:8261/api/v1/tasks/incremental_task_use_start_time_after_create/sources/mysql-01/migrate_targets
 	run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
 		"query-status $task_name" \
 		"\"stage\": \"Stopped\"" 2
@@ -1036,20 +1037,20 @@ function run() {
 	test_relay
 	test_source
 
-	test_shard_task
-	test_multi_tasks
-	test_noshard_task
-	test_task_templates
-	test_noshard_task_dump_status
-	test_complex_operations_of_source_and_task
-	test_task_with_ignore_check_items
-	test_delete_task_with_stopped_downstream
+	#test_shard_task
+	#test_multi_tasks
+	#test_noshard_task
+	#test_task_templates
+	#test_noshard_task_dump_status
+	#test_complex_operations_of_source_and_task
+	#test_task_with_ignore_check_items
+	#test_delete_task_with_stopped_downstream
 	test_start_task_with_condition
-	test_stop_task_with_condition
-	test_reverse_https
+	#test_stop_task_with_condition
+	#test_reverse_https
 
 	# NOTE: this test case MUST running at last, because it will offline some members of cluster
-	test_cluster
+	#test_cluster
 }
 
 cleanup_data openapi

--- a/dm/tests/openapi/run.sh
+++ b/dm/tests/openapi/run.sh
@@ -1037,20 +1037,20 @@ function run() {
 	test_relay
 	test_source
 
-	#test_shard_task
-	#test_multi_tasks
-	#test_noshard_task
-	#test_task_templates
-	#test_noshard_task_dump_status
-	#test_complex_operations_of_source_and_task
-	#test_task_with_ignore_check_items
-	#test_delete_task_with_stopped_downstream
+	test_shard_task
+	test_multi_tasks
+	test_noshard_task
+	test_task_templates
+	test_noshard_task_dump_status
+	test_complex_operations_of_source_and_task
+	test_task_with_ignore_check_items
+	test_delete_task_with_stopped_downstream
 	test_start_task_with_condition
-	#test_stop_task_with_condition
-	#test_reverse_https
+	test_stop_task_with_condition
+	test_reverse_https
 
 	# NOTE: this test case MUST running at last, because it will offline some members of cluster
-	#test_cluster
+	test_cluster
 }
 
 cleanup_data openapi

--- a/dm/worker/subtask.go
+++ b/dm/worker/subtask.go
@@ -652,8 +652,15 @@ func (st *SubTask) Update(ctx context.Context, cfg *config.SubTaskConfig) error 
 
 // OperateSchema operates schema for an upstream table.
 func (st *SubTask) OperateSchema(ctx context.Context, req *pb.OperateWorkerSchemaRequest) (schema string, err error) {
-	if !(st.Stage() == pb.Stage_Paused || st.stage == pb.Stage_Running && req.Op == pb.SchemaOp_ListMigrateTargets) {
-		return "", terror.ErrWorkerNotPausedStage.Generate(st.Stage().String())
+	switch req.Op {
+	case pb.SchemaOp_ListMigrateTargets:
+		if st.Stage() != pb.Stage_Running || st.Stage() != pb.Stage_Paused {
+			return "", terror.ErrWorkerNotPausedStage.Generate(st.Stage().String())
+		}
+	default:
+		if st.Stage() != pb.Stage_Paused {
+			return "", terror.ErrWorkerNotPausedStage.Generate(st.Stage().String())
+		}
 	}
 
 	syncUnit, ok := st.currUnit.(*syncer.Syncer)

--- a/dm/worker/subtask.go
+++ b/dm/worker/subtask.go
@@ -654,7 +654,7 @@ func (st *SubTask) Update(ctx context.Context, cfg *config.SubTaskConfig) error 
 func (st *SubTask) OperateSchema(ctx context.Context, req *pb.OperateWorkerSchemaRequest) (schema string, err error) {
 	switch req.Op {
 	case pb.SchemaOp_ListMigrateTargets:
-		if st.Stage() != pb.Stage_Running || st.Stage() != pb.Stage_Paused {
+		if st.Stage() != pb.Stage_Running && st.Stage() != pb.Stage_Paused {
 			return "", terror.ErrWorkerNotPausedStage.Generate(st.Stage().String())
 		}
 	default:

--- a/dm/worker/subtask.go
+++ b/dm/worker/subtask.go
@@ -664,7 +664,6 @@ func (st *SubTask) OperateSchema(ctx context.Context, req *pb.OperateWorkerSchem
 	if st.validatorStage() == pb.Stage_Running && req.Op != pb.SchemaOp_ListMigrateTargets {
 		return "", terror.ErrWorkerValidatorNotPaused.Generate(pb.Stage_Running.String())
 	}
-	log.L().Info("operate schema", zap.Stringer("operation", req.Op), zap.String("task", st.cfg.Name), zap.String("source", st.cfg.SourceID), zap.String("stage", st.Stage().String()), zap.String("unit stage", st.validatorStage().String()))
 
 	return syncUnit.OperateSchema(ctx, req)
 }

--- a/dm/worker/subtask.go
+++ b/dm/worker/subtask.go
@@ -652,7 +652,7 @@ func (st *SubTask) Update(ctx context.Context, cfg *config.SubTaskConfig) error 
 
 // OperateSchema operates schema for an upstream table.
 func (st *SubTask) OperateSchema(ctx context.Context, req *pb.OperateWorkerSchemaRequest) (schema string, err error) {
-	if st.Stage() != pb.Stage_Paused && req.Op != pb.SchemaOp_ListMigrateTargets {
+	if !(st.Stage() == pb.Stage_Paused || st.stage == pb.Stage_Running && req.Op == pb.SchemaOp_ListMigrateTargets) {
 		return "", terror.ErrWorkerNotPausedStage.Generate(st.Stage().String())
 	}
 
@@ -664,6 +664,7 @@ func (st *SubTask) OperateSchema(ctx context.Context, req *pb.OperateWorkerSchem
 	if st.validatorStage() == pb.Stage_Running && req.Op != pb.SchemaOp_ListMigrateTargets {
 		return "", terror.ErrWorkerValidatorNotPaused.Generate(pb.Stage_Running.String())
 	}
+	log.L().Info("operate schema", zap.Stringer("operation", req.Op), zap.String("task", st.cfg.Name), zap.String("source", st.cfg.SourceID), zap.String("stage", st.Stage().String()), zap.String("unit stage", st.validatorStage().String()))
 
 	return syncUnit.OperateSchema(ctx, req)
 }

